### PR TITLE
Added sudo user if using become method sudo for synchronize

### PR DESCRIPTION
**Pull request by cadavre**
##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

`synchronize`

##### ANSIBLE VERSION

`ansible 2.`

##### SUMMARY

If using `become=true` with `become_method=sudo` – take into consideration `become_user` if it was provided.

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure due to unit test errors. Here's one of them:

```
>           assert value, check
E           AssertionError: self.final_module_args['rsync_path'] == '"sudo rsync"'
E           assert False

test/units/plugins/action/test_synchronize.py:178: AssertionError
```

Full test results here: https://app.shippable.com/runs/58a5feccec8b7210004f9313/2/tests

**Comment by shoelzle** [comment]
Hello @cadavre ,

I guess you'll run into the same CI failures as me (https://app.shippable.com/runs/58a61b2ed0cccd0f006ca8a3/2/tests) and the test cases for the synchronize action plugin need to be adapted.

Also I noticed that there are CI failures for your fixes form 5071c80: https://app.shippable.com/runs/58a6cf9a119e6d0f0051b13d/2/tests
```
>                   if self._play_context.become_user not in ('root', remote_user):
E                   NameError: global name 'remote_user' is not defined
```

Since we started working on the same issue (#19746 and #21530), I'll stand by and wait. If I may help in any way, please let me know.

**Comment by jctanner** [comment]
cc @tima

**Comment by jctanner** [comment]
@cadavre @shoelzle please be very careful when you start modifying the action plugin and the tests. Don't just adapt the tests to have them pass the way you want them to. All of them were crafted with specific scenarios that we must continue to support and prevent from being broken by future commits. If you aren't sure, please hop on #ansible-devel on freenode and ask myself or abadger

**Comment by cadavre** [comment]
@jctanner existing tests were not even touched. Additional unit test for new scenario has been added though. Results for previous scenarios has not changed a bit. Also community proposed hack-fixes will still work after this PR. Solution is 100% backward compatible.

I would rather ask @abadger about implementation he made in https://github.com/ansible/ansible/commit/3cf59d30f72641ab51e0777e066cad1293295c67 which is very partial and not well researched.

File: lib/ansible/plugins/action/synchronize.py [hunk 1]
```diff
@@ -341,7 +341,10 @@ def run(self, tmp=None, task_vars=None):
                 # If no rsync_path is set, become was originally set, and dest is
                 # remote then add privilege escalation here.
                 if self._play_context.become_method == 'sudo':
-                    rsync_path = 'sudo rsync'
+                    if self._play_context.become_user not in ('root'):
+                        rsync_path = 'sudo -u %s rsync' % self._play_context.become_user
```
Comment (bcoca) [inline lib/ansible/plugins/action/synchronize.py]: just always set to this line, no need to make distinction on 'root', also some systems have alternateive 'root' users

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/action/synchronize.py (hunk 1 above)
Comment (cadavre) [reply lib/ansible/plugins/action/synchronize.py]: Ok, setting now to `self._play_context.become_user` but I guess tests won't cover non-"root" users under different systems.

**Review by cadavre — COMMENTED** [review]
Changed

**Comment by calfonso** [comment]
@cadavre We would like a slight modification to the implementation. We'd like you to provide an optional parameter that will toggle the behavior rather than providing a default. We want to ensure backwards compatibility. Note: this concern is based upon this being the synchronize plugin, which doesn't really conform to standards and can break modules in a number of ways.

**Comment by gundalow** [comment]
Hi,
This PR seems to have gotten stuck, so we are going to close it. We like to include community contributions, so this is a difficult call for us. 

Here's why we're closing this one:

* It's a feature
* We want to clear out the backlog so it's easier to find bug fixes
* There haven't been any comments or other activity since the PR was originally updated
* The source branch no longer exists in the PR authors repo

If you or someone else would like to see this change merged, please:
* Create a new PR and 
* Reference this author in the git commit & PR and
* Reference this PR in the new PR

How to rebase a branch: https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html
How to communicate with the Ansible community: https://docs.ansible.com/ansible/devel/community/communication.html
How to contribute to Ansible: https://docs.ansible.com/ansible/devel/community/index.html

**Pull request closed without merging**